### PR TITLE
docs: example of kg+vector index

### LINKED ISSUE #1318: RFC: Bring graph store to llama index
Dear Jerry/llama_index community!

May I propose changes to bring scalable graph capabilities to llama_index?

I had been watching llama_index for some time (impressed by how it's well abstracted, evaluated, and maintained) and would like to contribute. As one working around the graph, I had been thinking of similar things done by William(#433) and Logan(#487), and your discussions(around the embedding search of KG) in [the tread](https://twitter.com/jerryjliu0/status/1625531409137041409). 

- Due to the basic abstractions and PoC already designed and implemented, I could bring graph store on top of it, just like simple vector store v.s. other vector_stores. What do you think, please? If it's OK, I will prepare PR in this week, with the first graph_store implementation using NebulaGraph, the project I had been working on.

- Also, in the future, I am willing to discuss/contribute more around:
  - support KG + Vector index, multiple indices query as llama query context
  - NebulaGraph loader on llama hub(kg to doc)
  - Support indexing existing external graph store(i.e. adding an knowledge graph as index, create embedding to its entities and relationships)
  - vector store as KG embedding

The motivation is there are some cases in which we already have KG built from different sources(not just some docs), we could build smart services making use the most of the KG and LLAMA(index), and maybe we could leverage the (multiple hops)graph query to enable some knowledge not limited to the chunks of documents by adding the extra context on top of what embedding-vector search based magic could push this work to.

Thanks!

BR//Wey

---
Progress

- [x] Initial Impl. #2581 , blog: https://siwei.io/en/graph-enabled-llama-index/
- [ ] demo project
- [ ] docs/tutorial

### PR DESCRIPTION
**Pull request by wey-gu**
# Description

Part of #1318

- example of a custom retriever combining vector and kg, it's an example to load docs from [a new page in Wikipedia](https://en.wikipedia.org/wiki/2023_in_science) created after the 2021 Sept, where 1. how to best leverage kg + vector store backed custom index was demonstrated, 2. the extra cost, performance was evaluated. See rendered notebook [here](https://gist.github.com/wey-gu/93e1541d89657633649fb3f7afca3a9d).

- polishing kg store/index
  - now get_networkx_graph() will sample 100 starting nodes only(can be specified), as larger graph makes no sense to be rendered all
  - now the predicate of a triplet will be shown in the get_networkx_graph(), too
  - in some LLM models(i.e. azure oai gpt-3.5-turbo), the prompts extraction of knowledge won't ensure pure triplets being returned, extra logic to avoid invalid data was introduced
  - some optimization on the NebulaGraph graph store was introduced

Fixes #1318

PS, I am writing a blog post(and will record a video, too) together with this change's work.

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by jerryjliu — COMMENTED** [review]
For some reason I can't comment in-line in the notebook: 
- "we will extract more granular knowledges per each node" - could you be more precise here? e.g. "we will extract relationships between entities for each node" 

- In the "Define a CustomRetriever" section it might be good to add a description on what 1) the custom retriever does, and 2) how you think it can help augment results. 

- I really like the analysis section! 


Here's another high-level thought: right now you're doing retrieval from vector index and KG index independently, and combining them with a union (if mode == "OR"). Could it also make sense to do 1) retrieval from vector index first, and then 2) augment these retrieved set by traversing the KG graph for each node? I honestly don't know the right answer, curious to hear your take

**Review by logan-markewich — APPROVED** [review]
Thanks for contributing this notebook and analysis, as well as a few fixes in the KG index! 💪🏻 

Because of your notebook, I actually noticed a small bug with include_text in keyword retriever mode. I attached the fix to this PR :) 

Overall I think the analysis is pretty interesting. Was cool to see only ~150 tokens could be used, and llama-index still gives a good (but short) answer.

**Comment by wey-gu** [comment]
> Thanks for contributing this notebook and analysis, as well as a few fixes in the KG index! 💪🏻
> 
> Because of your notebook, I actually noticed a small bug with include_text in keyword retriever mode. I attached the fix to this PR :)
> 


wey: 

Dear @logan-markewich,
Many thanks for the optimization, review, great fix(sorry for introducing them), and supportive comments 🚀!

> Overall I think the analysis is pretty interesting. Was cool to see only ~150 tokens could be used, and llama-index still gives a good (but short) answer.

wey:

Yeah, if we combine some existing/general-kg, maybe in some use cases, LLM could even be considered to be decoupled, (or in a custom retriever, query KG first, and when it exceeds some threshold, to short-circuit other retrievers to reduce the context size)? XD

File: llama_index/indices/knowledge_graph/retriever.py [hunk 1]
```diff
@@ -147,6 +147,10 @@ def _retrieve(
                         # set. This helps introduce more global knowledge into the
                         # query. While it's more expensive, thus to be turned off
                         # by default, it can be useful for some applications.
+
+                        # TODO: we should a keyword-node_id map in IndexStruct, so that
```
Comment (wey-gu) [inline llama_index/indices/knowledge_graph/retriever.py:151]: When I am thinking of @jerryjliu 's idea and going through the existing retriever, I realized this is too costly when global explore is enabled and thus put this TODO here, @logan-markewich what do you think, please?

Comment (wey-gu) [reply llama_index/indices/knowledge_graph/retriever.py:151]: Sorry, I meant we already have KG.table(keyword->node_id), so just introduce a KG.<reverse_table>(node_id->keyword). in data_structs.py only adds a dict setting during `KG.add_node()`?

```python
    def add_node(self, keywords: List[str], node: Node) -> None:
        """Add text to table."""
        node_id = node.get_doc_id()
        for keyword in keywords:
            if keyword not in self.table:
                self.table[keyword] = set()
            self.table[keyword].add(node_id)
```

Comment (jerryjliu) [reply llama_index/indices/knowledge_graph/retriever.py:151]: @wey-gu i see, this makes sense. could you point me to the place where this doubles calls to the LLM for keyword extraction?

**Review by jerryjliu — APPROVED** [review]
nice!

File: llama_index/indices/knowledge_graph/retriever.py (hunk 1 above)
Comment (wey-gu) [reply llama_index/indices/knowledge_graph/retriever.py:151]: > @wey-gu i see, this makes sense. could you point me to the place where this doubles calls to the LLM for keyword extraction?

This line calls LLM to extract keywords from nodes( `self._get_keywords()`, in line 154 after this comment. By default, it's not called as `self.use_global_node_triplets` is off by default. :)

I can optimize this in the next PR.

```python
                        extended_subjs = self._get_keywords(
                            self._docstore.get_node(node_id).get_text()
                        )
```

**Comment by wey-gu** [comment]
Dear @jerryjliu 

> @wey-gu i see, this makes sense. could you point me to the place where this doubles calls to the LLM for keyword extraction?

This line calls LLM to extract keywords from nodes( `self._get_keywords()`, in line 154 after this comment. By default, it's not called as `self.use_global_node_triplets` is off by default. :)

I can optimize this in the next PR.

```python
                        extended_subjs = self._get_keywords(
                            self._docstore.get_node(node_id).get_text()
                        )
```

File: llama_index/graph_stores/nebulagraph.py [hunk 2]
```diff
@@ -327,7 +344,6 @@ def get_flat_rel_map(
         subjs_ = result.column_values("subj") or []
```
Comment (BleakStone) [inline llama_index/graph_stores/nebulagraph.py:344]: If the query keywords is not found in nebulagraph, the result type will be NoneType.
`result.column_values`  will cause 
AttributeError: 'NoneType' object has no attribute 'column_values'

Comment (BleakStone) [inline llama_index/graph_stores/nebulagraph.py:344]: If the query keywords is not found in nebulagraph, the result type will be NoneType.
result.column_values will cause
AttributeError: 'NoneType' object has no attribute 'column_values'

Comment (wey-gu) [reply llama_index/graph_stores/nebulagraph.py:344]: Thanks @BleakStone! I'll fix this in upcoming PRs, or if you are interested, you could help fix this, too!

**Comment by 123zzw** [comment]
hello，When I used this [example：](https://gptindex.readthedocs.io/en/latest/examples/query_engine/knowledge_graph_query_engine.html#)
I encountered this error, all of which are as follows：

> AttributeError: 'NoneType' object has no attribute 'CALL'

image


And the complete error is as follows：

> Generate NebulaGraph query from natural language.
> Use only the provided relationship types and properties in the schema.
> Do not use any other relationship types or properties that are not provided.
> Schema:
> ---
> Node properties: [{'tag': 'entity', 'properties': [('name', 'string')]}]
> Edge properties: [{'edge': 'relationship', 'properties': [('relationship', 'string')]}]
> Relationships: []
> 
> ---
> Note: NebulaGraph speaks a dialect of Cypher, comparing to standard Cypher:
> 
> 1. it uses double equals sign for comparison: `==` rather than `=`
> 2. it needs explicit label specification when referring to node properties, i.e.
> v is a variable of a node, and we know its label is Foo, v.`foo`.name is correct
> while v.name is not.
> 
> For example, see this diff between standard and NebulaGraph Cypher dialect:
> ```diff
> < MATCH (p:person)-[:directed]->(m:movie) WHERE m.name = 'The Godfather'
> < RETURN p.name;
> ---
> > MATCH (p:`person`)-[:directed]->(m:`movie`) WHERE m.`movie`.`name` == 'The Godfather'
> > RETURN p.`person`.`name`;
> ```
> 
> Question: Tell me about Peter Quill?
> 
> NebulaGraph Cypher dialect query:
> 
> Graph Store Query: ```css
> MATCH (p:Person)-[:ACTED_IN]->(m:Movie) WHERE m.name = 'The Godfather'
> RETURN p.name;
> ```
> This query will retrieve the name of the person who acted in the movie named 'The Godfather'.
> Traceback (most recent call last):
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/tenacity/__init__.py", line 382, in __call__
>     result = fn(*args, **kwargs)
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/graph_stores/nebulagraph.py", line 210, in execute
>     raise e
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/graph_stores/nebulagraph.py", line 186, in execute
>     raise ValueError(
> ValueError: Query failed. Query: ```css
> MATCH (p:Person)-[:ACTED_IN]->(m:Movie) WHERE m.name = 'The Godfather'
> RETURN p.name;
> ```
> This query will retrieve the name of the person who acted in the movie named 'The Godfather'., Param: {}Error message: SyntaxError: syntax error near ```css
> MA'
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/root/api_model/a_demotest/test/knowledgegraph_engine.py", line 102, in <module>
>     response = query_engine.query("Tell me about Peter Quill?")
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/indices/query/base.py", line 23, in query
>     response = self._query(str_or_query_bundle)
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/query_engine/knowledge_graph_query_engine.py", line 182, in _query
>     graph_store_response = self.graph_store.query(query=graph_store_query)
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/graph_stores/nebulagraph.py", line 526, in query
>     result = self.execute(query, param_map)
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/tenacity/__init__.py", line 289, in wrapped_f
>     return self(f, *args, **kw)
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/tenacity/__init__.py", line 379, in __call__
>     do = self.iter(retry_state=retry_state)
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/tenacity/__init__.py", line 326, in iter
>     raise retry_exc from fut.exception()
> tenacity.RetryError: RetryError[<Future at 0x7fe8a75d0a90 state=finished raised ValueError>]
> Exception ignored in: <function NebulaGraphStore.__del__ at 0x7fe8a7847700>
> Traceback (most recent call last):
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/graph_stores/nebulagraph.py", line 162, in __del__
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/nebula3/gclient/net/SessionPool.py", line 273, in close
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/nebula3/gclient/net/Session.py", line 288, in _sign_out
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/nebula3/gclient/net/Connection.py", line 197, in signout
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/nebula3/graph/GraphService.py", line 1603, in signout
>   File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/nebula3/graph/GraphService.py", line 1606, in send_signout
> AttributeError: 'NoneType' object has no attribute 'CALL'

**Comment by wey-gu** [comment]
Dear Sarah,

It’s strange I cannot find this thread from github, thus I am replying it via email.

It seems the text2cypher llm call returned abnormal output, may I know which model are you using as the LLM predictor, please?

From: Sarah Johnson ***@***.***>
Date: Tuesday, July 25, 2023 at 11:17
To: jerryjliu/llama_index ***@***.***>
Cc: Siwei Gu ***@***.***>, Mention ***@***.***>
Subject: Re: [jerryjliu/llama_index] docs: example of kg+vector index (PR #6497)

hello，When I used this example：<https://gptindex.readthedocs.io/en/latest/examples/query_engine/knowledge_graph_query_engine.html>
I encountered this error, all of which are as follows：

AttributeError: 'NoneType' object has no attribute 'CALL'

[image]<https://user-images.githubusercontent.com/65516153/255777833-21c898bf-103f-4b39-81c3-43148471a616.png>

And the complete error is as follows：

Generate NebulaGraph query from natural language.
Use only the provided relationship types and properties in the schema.
Do not use any other relationship types or properties that are not provided.
Schema:

Node properties: [{'tag': 'entity', 'properties': [('name', 'string')]}]
Edge properties: [{'edge': 'relationship', 'properties': [('relationship', 'string')]}]
Relationships: []

________________________________

Note: NebulaGraph speaks a dialect of Cypher, comparing to standard Cypher:

  1.  it uses double equals sign for comparison: == rather than =
  2.  it needs explicit label specification when referring to node properties, i.e.
v is a variable of a node, and we know its label is Foo, v.foo.name is correct
while v.name is not.

For example, see this diff between standard and NebulaGraph Cypher dialect:

< MATCH (p:person)-[:directed]->(m:movie) WHERE m.name = 'The Godfather'

< RETURN p.name;

---

> MATCH (p:`person`)-[:directed]->(m:`movie`) WHERE m.`movie`.`name` == 'The Godfather'

> RETURN p.`person`.`name`;

Question: Tell me about Peter Quill?

NebulaGraph Cypher dialect query:

Graph Store Query: ```css
MATCH (p:Person)-[:ACTED_IN]->(m:Movie) WHERE m.name = 'The Godfather'
RETURN p.name;

This query will retrieve the name of the person who acted in the movie named 'The Godfather'.

Traceback (most recent call last):

  File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/tenacity/__init__.py", line 382, in __call__

    result = fn(*args, **kwargs)

  File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/graph_stores/nebulagraph.py", line 210, in execute

    raise e

  File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/graph_stores/nebulagraph.py", line 186, in execute

    raise ValueError(

ValueError: Query failed. Query: ```css

MATCH (p:Person)-[:ACTED_IN]->(m:Movie) WHERE m.name = 'The Godfather'

RETURN p.name;

This query will retrieve the name of the person who acted in the movie named 'The Godfather'., Param: {}Error message: SyntaxError: syntax error near ```css
MA'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/root/api_model/a_demotest/test/knowledgegraph_engine.py", line 102, in
response = query_engine.query("Tell me about Peter Quill?")
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/indices/query/base.py", line 23, in query
response = self._query(str_or_query_bundle)
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/query_engine/knowledge_graph_query_engine.py", line 182, in _query
graph_store_response = self.graph_store.query(query=graph_store_query)
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/graph_stores/nebulagraph.py", line 526, in query
result = self.execute(query, param_map)
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/tenacity/init.py", line 289, in wrapped_f
return self(f, *args, **kw)
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/tenacity/init.py", line 379, in call
do = self.iter(retry_state=retry_state)
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/tenacity/init.py", line 326, in iter
raise retry_exc from fut.exception()
tenacity.RetryError: RetryError[<Future at 0x7fe8a75d0a90 state=finished raised ValueError>]
Exception ignored in: <function NebulaGraphStore.del at 0x7fe8a7847700>
Traceback (most recent call last):
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/llama_index/graph_stores/nebulagraph.py", line 162, in del
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/nebula3/gclient/net/SessionPool.py", line 273, in close
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/nebula3/gclient/net/Session.py", line 288, in _sign_out
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/nebula3/gclient/net/Connection.py", line 197, in signout
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/nebula3/graph/GraphService.py", line 1603, in signout
File "/root/miniconda3/envs/llm-test/lib/python3.9/site-packages/nebula3/graph/GraphService.py", line 1606, in send_signout
AttributeError: 'NoneType' object has no attribute 'CALL'

**Pull request merged**